test/system/src/test/java/io/pravega/test/system/SegmentReaderAPITest.java

```diff
@@ -0,0 +1,517 @@
+import io.pravega.client.stream.impl.ClientFactoryImpl;
+import io.pravega.client.stream.impl.StreamCutImpl;
+import io.pravega.client.stream.impl.UTF8StringSerializer;
+import io.pravega.common.concurrent.ExecutorServiceHelpers;
+import io.pravega.common.hash.RandomFactory;
+import io.pravega.test.system.framework.Environment;
+import io.pravega.test.system.framework.SystemTestRunner;
+import io.pravega.test.system.framework.Utils;
+import io.pravega.test.system.framework.services.Service;
+import lombok.Cleanup;
+import lombok.extern.slf4j.Slf4j;
+import mesosphere.marathon.client.MarathonException;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import java.net.URI;
+import java.util.Random;
+import java.util.UUID;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Collections;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static io.pravega.shared.NameUtils.computeSegmentId;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertEquals;
+
+@Slf4j
+@RunWith(SystemTestRunner.class)
+public class SegmentReaderAPITest extends AbstractReadWriteTest {
+    private static final String DATA_OF_SIZE_30 = "this is a test strings"; // data length = 22 bytes , header = 8 bytes
+    private final Random random = RandomFactory.create();
+    private URI controllerURI = null;
+    private StreamManager streamManager = null;
+    private Controller controller = null;
+    private ConnectionFactory connectionFactory = null;
+    private final ScheduledExecutorService executor = ExecutorServiceHelpers.newScheduledThreadPool(1, "executor");
+
+    /**
+     * This is used to setup the various services required by the system test framework.
+     */
+    @Environment
+    public static void initialize() throws MarathonException {
+        URI zkUri = startZookeeperInstance();
+        startBookkeeperInstances(zkUri);
+        URI controllerUri = ensureControllerRunning(zkUri);
+        ensureSegmentStoreRunning(zkUri, controllerUri);
+    }
+
+    @Before
+    public void setup() {
+        controllerURI = fetchControllerURI();
+        streamManager = StreamManager.create(Utils.buildClientConfig(controllerURI));
+
+        connectionFactory = new SocketConnectionFactoryImpl(ClientConfig.builder().build());
+        controller = new ControllerImpl(ControllerImplConfig.builder()
+                .clientConfig(Utils.buildClientConfig(controllerURI))
+                .build(), connectionFactory.getInternalExecutor());
+    }
+
+    @After
+    public void tearDown() {
+        streamManager.close();
+        controller.close();
+        connectionFactory.close();
```

Comment:
connectionFactory is an internal client objects there is no reason to manually manage their lifecycle here. StreamManager can move into the test like ClientFactory. Same for controller.